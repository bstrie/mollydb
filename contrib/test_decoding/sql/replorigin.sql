-- predictability
SET synchronous_commit = on;

CREATE TABLE origin_tbl(id serial primary key, data text);
CREATE TABLE target_tbl(id serial primary key, data text);

SELECT mdb_replication_origin_create('test_decoding: regression_slot');
-- ensure duplicate creations fail
SELECT mdb_replication_origin_create('test_decoding: regression_slot');

--ensure deletions work (once)
SELECT mdb_replication_origin_create('test_decoding: temp');
SELECT mdb_replication_origin_drop('test_decoding: temp');
SELECT mdb_replication_origin_drop('test_decoding: temp');

SELECT 'init' FROM mdb_create_logical_replication_slot('regression_slot', 'test_decoding');

-- origin tx
INSERT INTO origin_tbl(data) VALUES ('will be replicated and decoded and decoded again');
INSERT INTO target_tbl(data)
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1');

-- as is normal, the insert into target_tbl shows up
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1');

INSERT INTO origin_tbl(data) VALUES ('will be replicated, but not decoded again');

-- mark session as replaying
SELECT mdb_replication_origin_session_setup('test_decoding: regression_slot');

-- ensure we prevent duplicate setup
SELECT mdb_replication_origin_session_setup('test_decoding: regression_slot');

SELECT '' FROM mdb_logical_emit_message(false, 'test', 'this message will not be decoded');

BEGIN;
-- setup transaction origin
SELECT mdb_replication_origin_xact_setup('0/aabbccdd', '2013-01-01 00:00');
INSERT INTO target_tbl(data)
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1', 'only-local', '1');
COMMIT;

-- check replication progress for the session is correct
SELECT mdb_replication_origin_session_progress(false);
SELECT mdb_replication_origin_session_progress(true);

SELECT mdb_replication_origin_session_reset();

SELECT local_id, external_id, remote_lsn, local_lsn <> '0/0' FROM mdb_replication_origin_status;

-- check replication progress identified by name is correct
SELECT mdb_replication_origin_progress('test_decoding: regression_slot', false);
SELECT mdb_replication_origin_progress('test_decoding: regression_slot', true);

-- ensure reset requires previously setup state
SELECT mdb_replication_origin_session_reset();

-- and magically the replayed xact will be filtered!
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1', 'only-local', '1');

--but new original changes still show up
INSERT INTO origin_tbl(data) VALUES ('will be replicated');
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1',  'only-local', '1');

SELECT mdb_drop_replication_slot('regression_slot');
SELECT mdb_replication_origin_drop('test_decoding: regression_slot');
