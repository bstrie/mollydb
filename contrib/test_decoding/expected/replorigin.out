-- predictability
SET synchronous_commit = on;
CREATE TABLE origin_tbl(id serial primary key, data text);
CREATE TABLE target_tbl(id serial primary key, data text);
SELECT mdb_replication_origin_create('test_decoding: regression_slot');
 mdb_replication_origin_create 
------------------------------
                            1
(1 row)

-- ensure duplicate creations fail
SELECT mdb_replication_origin_create('test_decoding: regression_slot');
ERROR:  duplicate key value violates unique constraint "mdb_replication_origin_roname_index"
DETAIL:  Key (roname)=(test_decoding: regression_slot) already exists.
--ensure deletions work (once)
SELECT mdb_replication_origin_create('test_decoding: temp');
 mdb_replication_origin_create 
------------------------------
                            2
(1 row)

SELECT mdb_replication_origin_drop('test_decoding: temp');
 mdb_replication_origin_drop 
----------------------------
 
(1 row)

SELECT mdb_replication_origin_drop('test_decoding: temp');
ERROR:  cache lookup failed for replication origin 'test_decoding: temp'
SELECT 'init' FROM mdb_create_logical_replication_slot('regression_slot', 'test_decoding');
 ?column? 
----------
 init
(1 row)

-- origin tx
INSERT INTO origin_tbl(data) VALUES ('will be replicated and decoded and decoded again');
INSERT INTO target_tbl(data)
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1');
-- as is normal, the insert into target_tbl shows up
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1');
                                                                                    data                                                                                    
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 BEGIN
 table public.target_tbl: INSERT: id[integer]:1 data[text]:'BEGIN'
 table public.target_tbl: INSERT: id[integer]:2 data[text]:'table public.origin_tbl: INSERT: id[integer]:1 data[text]:''will be replicated and decoded and decoded again'''
 table public.target_tbl: INSERT: id[integer]:3 data[text]:'COMMIT'
 COMMIT
(5 rows)

INSERT INTO origin_tbl(data) VALUES ('will be replicated, but not decoded again');
-- mark session as replaying
SELECT mdb_replication_origin_session_setup('test_decoding: regression_slot');
 mdb_replication_origin_session_setup 
-------------------------------------
 
(1 row)

-- ensure we prevent duplicate setup
SELECT mdb_replication_origin_session_setup('test_decoding: regression_slot');
ERROR:  cannot setup replication origin when one is already setup
SELECT '' FROM mdb_logical_emit_message(false, 'test', 'this message will not be decoded');
 ?column? 
----------
 
(1 row)

BEGIN;
-- setup transaction origin
SELECT mdb_replication_origin_xact_setup('0/aabbccdd', '2013-01-01 00:00');
 mdb_replication_origin_xact_setup 
----------------------------------
 
(1 row)

INSERT INTO target_tbl(data)
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1', 'only-local', '1');
COMMIT;
-- check replication progress for the session is correct
SELECT mdb_replication_origin_session_progress(false);
 mdb_replication_origin_session_progress 
----------------------------------------
 0/AABBCCDD
(1 row)

SELECT mdb_replication_origin_session_progress(true);
 mdb_replication_origin_session_progress 
----------------------------------------
 0/AABBCCDD
(1 row)

SELECT mdb_replication_origin_session_reset();
 mdb_replication_origin_session_reset 
-------------------------------------
 
(1 row)

SELECT local_id, external_id, remote_lsn, local_lsn <> '0/0' FROM mdb_replication_origin_status;
 local_id |          external_id           | remote_lsn | ?column? 
----------+--------------------------------+------------+----------
        1 | test_decoding: regression_slot | 0/AABBCCDD | t
(1 row)

-- check replication progress identified by name is correct
SELECT mdb_replication_origin_progress('test_decoding: regression_slot', false);
 mdb_replication_origin_progress 
--------------------------------
 0/AABBCCDD
(1 row)

SELECT mdb_replication_origin_progress('test_decoding: regression_slot', true);
 mdb_replication_origin_progress 
--------------------------------
 0/AABBCCDD
(1 row)

-- ensure reset requires previously setup state
SELECT mdb_replication_origin_session_reset();
ERROR:  no replication origin is configured
-- and magically the replayed xact will be filtered!
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1', 'only-local', '1');
 data 
------
(0 rows)

--but new original changes still show up
INSERT INTO origin_tbl(data) VALUES ('will be replicated');
SELECT data FROM mdb_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1',  'only-local', '1');
                                      data                                      
--------------------------------------------------------------------------------
 BEGIN
 table public.origin_tbl: INSERT: id[integer]:3 data[text]:'will be replicated'
 COMMIT
(3 rows)

SELECT mdb_drop_replication_slot('regression_slot');
 mdb_drop_replication_slot 
--------------------------
 
(1 row)

SELECT mdb_replication_origin_drop('test_decoding: regression_slot');
 mdb_replication_origin_drop 
----------------------------
 
(1 row)

